src/dump.c

```diff
@@ -1821,6 +1839,59 @@ static jl_value_t *jl_deserialize_value_singleton(jl_serializer_state *s, jl_val
     return v;
 }
 
+static void jl_deserialize_struct(jl_serializer_state *s, jl_value_t *v, size_t startfield)
+{
+    jl_datatype_t *dt = (jl_datatype_t*)jl_typeof(v);
+    size_t i, nf = jl_datatype_nfields(dt);
+    char *data = (char*)jl_data_ptr(v);
+    for (i = startfield; i < nf; i++) {
+        if (jl_field_size(dt, i) > 0) {
+            if (jl_field_isptr(dt, i)) {
+                jl_value_t **fld = (jl_value_t**)(data+jl_field_offset(dt, i));
+                *fld = jl_deserialize_value(s, fld);
+            }
+            else {
+                jl_set_nth_field(v, i, jl_deserialize_value(s, NULL));
+            }
+        }
+    }
+    if (s->mode == MODE_MODULE) {
+        if (dt == jl_typename_type) {
+            jl_typename_t *tn = (jl_typename_t*)v;
+            tn->cache = jl_emptysvec; // the cache is refilled later (tag 5)
+            tn->linearcache = jl_emptysvec; // the cache is refilled later (tag 5)
+        }
+        if (dt == jl_typemap_entry_type) {
+            if (((jl_typemap_entry_t*)v)->max_world == ~(size_t)0) {
+                // update world validity to reflect current state of the counter
+                ((jl_typemap_entry_t*)v)->min_world = jl_world_counter;
+            }
+            else {
+                // garbage entry - delete it :(
+                ((jl_typemap_entry_t*)v)->min_world = ((jl_typemap_entry_t*)v)->max_world - 1;
+            }
+        }
+    }
+}
+
+static jl_value_t *jl_deserialize_typemap_entry(jl_serializer_state *s)
+{
+    int N = read_int32(s->s); int n = N;
+    jl_value_t *te = jl_nothing;
+    jl_value_t **pn = &te;
+    while (n > 0) {
+        jl_value_t *v = jl_gc_alloc(s->ptls, jl_datatype_size(jl_typemap_entry_type), jl_typemap_entry_type);
+        if (n == N && s->mode != MODE_AST)
+            arraylist_push(&backref_list, v);
+        jl_deserialize_struct(s, v, 1);
+        ((jl_typemap_entry_t*)v)->next = (jl_typemap_entry_t*)jl_nothing;
+        *pn = v;
```

Comment:
Yes, so if you are sure that the parent is always allocated in the same "deserialization session" it should be enough with a comment about it. Looks like it should be fine here since `*pn` is pointing to either the stack or a `v` allocated in a previous iteration?